code/goonstation/modules/power/power_transmission_laser/transmission_laser.dm

```diff
@@ -0,0 +1,524 @@
+
+#define A1_CURVE 20
+
+/obj/machinery/power/transmission_laser/proc/sell_power(joules)
+	var/mega_joules = joules / (1 MW)
+
+	var/generated_cash = (2 * mega_joules * PROCESS_CAP) / ((2 * mega_joules) + (PROCESS_CAP * A1_CURVE))
+	generated_cash += (4 * mega_joules * MINIMUM_BAR) / (4 * mega_joules + MINIMUM_BAR)
+	generated_cash = round(generated_cash)
+	if(generated_cash < 0)
+		return
+
+	total_energy += joules
+	total_earnings += generated_cash
+	generated_cash += unsent_earnings
+	unsent_earnings = generated_cash
+
+	var/datum/money_account/engineering_bank_account = GLOB.station_money_database.get_account_by_department(DEPARTMENT_ENGINEERING)
+	var/datum/money_account/cargo_bank_account = GLOB.station_money_database.get_account_by_department(DEPARTMENT_SUPPLY)
+
+	var/medium_cut = generated_cash * 0.25
+	var/high_cut = generated_cash * 0.75
+
+	cargo_bank_account.deposit_credits(medium_cut, "Transmission Laser Payout")
+	unsent_earnings -= medium_cut
+
+	engineering_bank_account.deposit_credits(high_cut, "Transmission Laser Payout")
+	unsent_earnings -= high_cut
+
+#undef A1_CURVE
+#undef PROCESS_CAP
+#undef MINIMUM_BAR
+
+// Beam related procs
+
+/obj/machinery/power/transmission_laser/proc/setup_lasers()
+	var/turf/last_step = get_step(get_front_turf(), dir)
+	// Create new lasers from the starting point to either the blocker or the edge of the map
+	for(var/num = 1 to range + 1)
+		if(!(locate(/obj/effect/transmission_beam/) in last_step))
+			var/obj/effect/transmission_beam/new_beam = new(last_step, src)
+			new_beam.host = src
+			new_beam.dir = dir
+			laser_effects += new_beam
+
+		last_step = get_step(last_step, dir)
+
+/obj/machinery/power/transmission_laser/proc/destroy_lasers()
+	for(var/obj/effect/transmission_beam/listed_beam as anything in laser_effects)
+		laser_effects -= listed_beam
+		qdel(listed_beam)
+
+/obj/machinery/power/transmission_laser/proc/shorten_beam()
+	for(var/obj/effect/transmission_beam/listed_beam as anything in laser_effects)
+		if(get_dist(get_front_turf(), listed_beam.loc) > range)
+			laser_effects -= listed_beam
+			qdel(listed_beam)
+
+/// Affect the atom according to it's type and the ouput power of the laser.
+/obj/machinery/power/transmission_laser/proc/atom_beam_effect(atom/beam_target)
+	if(QDELETED(beam_target) || istype(beam_target, /obj/structure/window))
+		return
+	var/mw_power = (output_number * power_format_multi_output) / (1 MW)
+	if(isliving(beam_target) && !istype(beam_target, /mob/living/simple_animal/revenant))
+		var/mob/living/victim = beam_target
+		switch(mw_power)
+			if(0 to 25)
+				victim.adjustFireLoss(-mw_power * 15)
+				victim.adjust_fire_stacks(mw_power)
+			if(26 to 50)
+				victim.gib(FALSE)
+			else
+				explosion(victim, 3, 2, 2)
+				victim.gib(FALSE)
+
+	else if(istype(beam_target,/obj/))
+		var/obj/target_object = beam_target
+		target_object.take_damage(mw_power)
+	else if(istype(beam_target, /turf/simulated/wall)) // We don't want to damage floors
```

Comment:
pretty sure we have a define for this like `iswallturf()`

Reply:
done